pip/req/req_set.py

```diff
@@ -267,9 +271,12 @@ def add_requirement(self, install_req, parent_req_name=None):
             return result
 
     def has_requirement(self, project_name):
-        for name in project_name, project_name.lower():
-            if name in self.requirements or name in self.requirement_aliases:
-                return True
+        name = project_name.lower()
```

Comment:
Yes, without this pip will never attempt to process the package, and we'll never return to add_requirement, silently dropping the package.

Also from what I can gather, not filtering constraints from this check we're changing the meaning of the has_requirement function from "we have this requirement and we've processed it, don't worry it's been processed" to "we either have a constraint that won't be installed, or we have a package that has actually been processed".